APISyncExternalModule.php

```diff
@@ -898,25 +905,291 @@ function cancelSync(){
+			foreach ($translations as $i => $form_names) {
+				// [redcap_repeat_instrument] component
+				$local_form_name = $form_names[0];
+				$export_form_name = $form_names[1];
+				if ($proj_prefix == 'export-') {
+					// export logic
+					if ($instance['redcap_repeat_instrument'] == $local_form_name) {
+						$instance['redcap_repeat_instrument'] = $export_form_name;
+					}
+					
+					// [$form . '_complete'] components
+					if (isset($instance[$local_form_name . '_complete'])) {
+						$instance[$export_form_name . '_complete'] = $instance[$local_form_name . '_complete'];
+						unset($instance[$local_form_name . '_complete']);
+					}
+				} else {
+					// import logic
+					if (in_array($instance['redcap_repeat_instrument'], $form_names, true)) {
+						$instance['redcap_repeat_instrument'] = $local_form_name;
+					}
+					
+					// [$form . '_complete'] components
+					foreach($form_names as $other_form_name) {
+						if (isset($instance[$other_form_name . '_complete'])) {
+							$instance[$local_form_name . '_complete'] = $instance[$other_form_name . '_complete'];
+							unset($instance[$other_form_name . '_complete']);
+						}
+					}
+				}
+			}
+		}
+	}
+	
+	public function translateEventNames(&$data, &$project) {
+		$proj_prefix = $this->getProjectTypePrefix($project);
+		$translations = $project[$proj_prefix . 'event-translations'];
+		if (gettype($translations) != 'array') {
+			return;
+		}
+		// translate [redcap_event_name] fields where applicable
+		foreach ($data as &$instance) {
+			foreach ($translations as $event_names) {
+				// [redcap_event_name] component
+				$local_event_name = $event_names[0];
+				$export_event_name = $event_names[1];
+				$event_name_pieces = explode('_arm_', $instance['redcap_event_name']);
+				$event_name = $event_name_pieces[0];
+				$event_arm_number = $event_name_pieces[1];
+				if ($proj_prefix == 'export-') {
+					// export logic
+					if ($event_name == $local_event_name) {
+						$instance['redcap_event_name'] = $export_event_name . "_arm_" . $event_arm_number;
+						break;
+					}
+				} else {
+					// import logic
+					if (in_array($event_name, $event_names, true)) {
+						$instance['redcap_event_name'] = $local_event_name . "_arm_" . $event_arm_number;
+						break;
+					}
+				}
+			}
+		}
+	}
+	
+	public function buildTranslations(&$project) {
+		// function will only build translations if json present in $project['form-translations'/'event-translations']
+		if ($this->translationsAreBuilt($project)) {
+			return;
+		}
+		$proj_key_prefix = $this->getProjectTypePrefix($project);
+		
+		foreach (['form', 'event'] as $type) {
+			$setting = &$project[$proj_key_prefix . "$type-translations"];
+			$setting = json_decode($setting, true);
+			if ($setting) {
+				foreach($setting as $i => $row) {
+					foreach($row as $j => $name) {
+						$func_name = "format" . ucfirst($type) . "Name";
+						$setting[$i][$j] = $this->$func_name($name);
+					}
+				}
+			}
+		}
+	}
```

Comment:
Instead of this setting modification step, would it be simpler just to call `formatFormName()` in the appropriate place in `translateFormNames()`, and 'formatEventName()` in the appropriate place in `translateEventNames()`?

Reply:
Hm. I think it's simpler to format in buildTranslations. My reasoning being that in translateFormNames, I have

`if (in_array($instance['redcap_repeat_instrument'], $form_names, true)) {`

So if I were to format in the translation functions, I'd have to create a separate loop to format names before I translate the data. Which sounds more like building translations than applying them, to me, anyhow.

That being said, if you have a preference for formatting inside of the translate functions, it'd be easy for me to implement, just let me know.

Reply:
I do suspect there is a way to do the same thing with a lot less lines of code, but I'm just going to resolve this since the quote is pretty much exhausted.